File: samples/AdventuresWorksLT.VisualBasic/AdventureWorksLT/BaseModel.vb
<UniqueConstraint("AK_RowGuid", DbName:="AK_%_rowguid", Description:="Unique nonclustered constraint. Used to support replication samples.")>
Public MustInherit Class BaseModel(Of T As CandidateKey)
    Inherits Model(Of T)

    Shared Sub New()
        RegisterColumn(Function(x As BaseModel(Of T)) x.RowGuid)
        RegisterColumn(Function(x As BaseModel(Of T)) x.ModifiedDate)
    End Sub

    Private m_RowGuid As _Guid
    <Required>
    <AutoGuid(Name:="DF_%_rowguid", Description:="Default constraint value of NEWID()")>
    <DbColumn(Description:="ROWGUIDCOL number uniquely identifying the record. Used to support a merge replication sample.")>
    Public Property RowGuid As _Guid
        Get
            Return m_RowGuid
        End Get
        Private Set
            m_RowGuid = Value
        End Set
    End Property

    Private _ModifiedDate As _DateTime
    <Required>
    <SqlDateTime>
    <AutoDateTime(Name:="DF_%_ModifiedDate", Description:="Default constraint value of GETDATE()")>
    <DbColumn(Description:="Date and time the record was last updated.")>
    Public Property ModifiedDate As _DateTime
        Get
            Return _ModifiedDate
        End Get
        Private Set
            _ModifiedDate = Value
        End Set
    End Property

    Public Sub ResetRowIdentifiers()
        Dim vDataSet = DataSet
        If vDataSet Is Nothing Then Return

        For i As Integer = 0 To vDataSet.Count - 1
            RowGuid(i) = Guid.NewGuid()
            ModifiedDate(i) = DateTime.Now
        Next
    End Sub

    <_UniqueConstraint>
    Private ReadOnly Property AK_RowGuid As ColumnSort()
        Get
            Return New ColumnSort() {RowGuid}
        End Get
    End Property
End Class

File: samples/AdventureWorksLT.Data.SqlServer.CSharp/AdventureWorksLT.Data/BaseModel.cs
using DevZest.Data;
using DevZest.Data.Annotations;
using DevZest.Data.SqlServer;
using System;

namespace DevZest.Samples.AdventureWorksLT
{
    [UniqueConstraint(nameof(AK_RowGuid), DbName = "AK_%_rowguid", Description = "Unique nonclustered constraint. Used to support replication samples.")]
    public abstract class BaseModel<T> : Model<T>
        where T : CandidateKey
    {
        static BaseModel()
        {
            RegisterColumn((BaseModel<T> x) => x.RowGuid);
            RegisterColumn((BaseModel<T> x) => x.ModifiedDate);
        }

        [Required]
        [AutoGuid(Name = "DF_%_rowguid", Description = "Default constraint value of NEWID()")]
        [DbColumn(Description = "ROWGUIDCOL number uniquely identifying the record. Used to support a merge replication sample.")]
        public _Guid RowGuid { get; private set; }

        [Required]
        [SqlDateTime]
        [AutoDateTime(Name = "DF_%_ModifiedDate", Description = "Default constraint value of GETDATE()")]
        [DbColumn(Description = "Date and time the record was last updated.")]
        public _DateTime ModifiedDate { get; private set; }

        public void ResetRowIdentifiers()
        {
            var dataSet = DataSet;
            if (dataSet == null)
                return;
            for (int i = 0; i < dataSet.Count; i++)
            {
                RowGuid[i] = Guid.NewGuid();
                ModifiedDate[i] = DateTime.Now;
            }
        }

        [_UniqueConstraint]
        private ColumnSort[] AK_RowGuid
        {
            get { return new ColumnSort[] { RowGuid }; }
        }
    }
}
